File: go/engine_test.go
package amino_test

import (
	"errors"
	"reflect"
	"testing"

	amino "github.com/raiderrobert/amino/go"
)

func mustEngine(t *testing.T, schema string, opts ...amino.Option) *amino.Engine {
	t.Helper()
	e, err := amino.LoadSchema(schema, opts...)
	if err != nil {
		t.Fatal(err)
	}
	return e
}

func code(err error) amino.Code {
	var ae *amino.Error
	if errors.As(err, &ae) {
		return ae.Code
	}
	return ""
}

func TestSchemaParseAndExport(t *testing.T) {
	src := `# comment
struct Address { street: Str, city: Str }
struct Customer {
    id: Str
    billing: Address
    shipping: Address?
}
amount: Int {min: 0, max: 100}
tags: List[Str|Int]
email: Str?
customer: Customer
score: (a: Int, b: Float?) -> Float
`
	e := mustEngine(t, src)
	s := e.Schema()
	if len(s.Structs) != 2 || len(s.Fields) != 4 || len(s.Functions) != 1 {
		t.Fatalf("counts: %d structs %d fields %d functions", len(s.Structs), len(s.Fields), len(s.Functions))
	}
	if f := s.Lookup("customer.billing.city"); f == nil || f.TypeName != "Str" {
		t.Fatalf("nested lookup: %+v", f)
	}
	if f := s.Lookup("tags"); f.Kind != amino.KindList || !reflect.DeepEqual(f.ElementTypes, []string{"Str", "Int"}) {
		t.Fatalf("list: %+v", f)
	}
	if f := s.Lookup("amount"); f.Constraints["min"] != int64(0) {
		t.Fatalf("constraints: %+v", f.Constraints)
	}
	want := "struct Address {street: Str, city: Str}\nstruct Customer {id: Str, billing: Address, shipping: Address?}\namount: Int {max: 100, min: 0}\ntags: List[Str|Int]\nemail: Str?\ncustomer: Customer\nscore: (a: Int, b: Float?) -> Float"
	if got := e.ExportSchema(); got != want {
		t.Fatalf("export:\n%s\nwant:\n%s", got, want)
	}
}

func TestSchemaValidation(t *testing.T) {
	cases := map[string]string{
		"a: Nope":                              "unknown type",
		"a: Int\na: Str":                       "duplicate",
		"struct A { b: B }\nstruct B { a: A }": "circular",
		"struct S { x: Int, x: Str }":          "duplicate field",
		"a Int":                                "expected",
		"struct: Int":                          "reserved",
	}
	for src, want := range cases {
		_, err := amino.LoadSchema(src)
		if err == nil {
			t.Errorf("%q: expected error containing %q", src, want)
			continue
		}
		if c := code(err); c != amino.CodeSchemaParse && c != amino.CodeSchemaValidation {
			t.Errorf("%q: code %s", src, c)
		}
	}
}

func TestEvalAndMatchModes(t *testing.T) {
	e := mustEngine(t, "amount: Int\nstate: Str")
	rules := []amino.Rule{
		{ID: "1", Expr: "amount > 0 and state = 'CA'", Meta: map[string]any{"ordering": 3}},
		{ID: "2", Expr: "amount > 10 and state = 'CA'", Meta: map[string]any{"ordering": 2}},
		{ID: "3", Expr: "amount >= 100", Meta: map[string]any{"ordering": 1}},
	}
	rec := map[string]any{"id": 45, "amount": 100, "state": "CA"}

	r, err := e.Eval(rules, rec, nil)
	if err != nil {
		t.Fatal(err)
	}
	if !reflect.DeepEqual(r.Matched, []string{"1", "2", "3"}) || r.ID != 45 {
		t.Fatalf("all: %+v", r)
	}

	r, _ = e.Eval(rules, rec, &amino.MatchConfig{Mode: "first", Key: "ordering", Order: "asc"})
	if !reflect.DeepEqual(r.Matched, []string{"3"}) {
		t.Fatalf("first asc: %+v", r)
	}
	r, _ = e.Eval(rules, rec, &amino.MatchConfig{Mode: "first", Key: "ordering", Order: "desc"})
	if !reflect.DeepEqual(r.Matched, []string{"1"}) {
		t.Fatalf("first desc: %+v", r)
	}

	r, _ = e.Eval(rules, map[string]any{"amount": 5, "state": "NY"}, &amino.MatchConfig{Mode: "inverse"})
	if !reflect.DeepEqual(r.Excluded, []string{"1", "2", "3"}) {
		t.Fatalf("inverse: %+v", r)
	}

	scoring := []amino.Rule{{ID: "a", Expr: "amount"}, {ID: "b", Expr: "amount > 0"}}
	th := 50.0
	r, _ = e.Eval(scoring, rec, &amino.MatchConfig{Mode: "score", Threshold: &th})
	if r.Score == nil || *r.Score != 101 || !reflect.DeepEqual(r.Matched, []string{"a", "b"}) {
		t.Fatalf("score: %+v", r)
	}
}

func TestMissingFieldMakesRuleFalse(t *testing.T) {
	e := mustEngine(t, "a: Int\nb: Int?")
	r, _ := e.Eval([]amino.Rule{{ID: "r", Expr: "not b = 1"}, {ID: "s", Expr: "b = 1 or a = 1"}}, map[string]any{"a": 1}, nil)
	if !reflect.DeepEqual(r.Matched, []string{"s"}) {
		t.Fatalf("got %+v", r)
	}
}

func TestDecisionValidation(t *testing.T) {
	strict := mustEngine(t, "n: Int {min: 0}\ncontact: email", amino.WithDecisionsMode("strict"))
	_, err := strict.Eval([]amino.Rule{{ID: "r", Expr: "n > 1"}}, map[string]any{"n": "x", "contact": "a@b.co"}, nil)
	if code(err) != amino.CodeDecisionValidation {
		t.Fatalf("strict type: %v", err)
	}
	_, err = strict.Eval([]amino.Rule{{ID: "r", Expr: "n > 1"}}, map[string]any{"n": -1, "contact": "a@b.co"}, nil)
	if code(err) != amino.CodeDecisionValidation {
		t.Fatalf("strict constraint: %v", err)
	}
	_, err = strict.Eval([]amino.Rule{{ID: "r", Expr: "n > 1"}}, map[string]any{"n": 2, "contact": "nope"}, nil)
	if code(err) != amino.CodeDecisionValidation {
		t.Fatalf("strict custom type: %v", err)
	}

	loose := mustEngine(t, "n: Int\ncontact: email")
	r, err := loose.Eval([]amino.Rule{{ID: "r", Expr: "contact = 'nope'"}}, map[string]any{"n": 2, "contact": "nope"}, nil)
	if err != nil || len(r.Matched) != 0 || len(r.Warnings) != 1 {
		t.Fatalf("loose: %+v %v", r, err)
	}
}

func TestFunctionsAndCustomOperators(t *testing.T) {
	e := mustEngine(t, "text: Str\ntoxicity: (t: Str) -> Float\na: Int\nb: Int",
		amino.WithFunction("toxicity", func(args ...any) (any, error) { return 0.9, nil }),
		amino.WithOperator(amino.OperatorDef{
			Keyword: "near", BindingPower: 40, InputTypes: []string{"Int", "Int"}, ReturnType: "Bool",
			Fn: func(args ...any) (any, error) {
				x, _ := args[0].(int64)
				y, _ := args[1].(int64)
				d := x - y
				if d < 0 {
					d = -d
				}
				return d < 5, nil
			},
		}),
	)
	r, err := e.Eval([]amino.Rule{{ID: "t", Expr: "toxicity(text) > 0.5"}, {ID: "n", Expr: "a near b"}},
		map[string]any{"text": "hi", "a": 3, "b": 6}, nil)
	if err != nil {
		t.Fatal(err)
	}
	if !reflect.DeepEqual(r.Matched, []string{"t", "n"}) {
		t.Fatalf("got %+v", r)
	}
	if _, err := e.Parse("toxicity(a) > 0.5"); code(err) != amino.CodeTypeMismatch {
		t.Fatalf("argument type: %v", err)
	}
	if _, err := e.Parse("toxicity()"); code(err) != amino.CodeTypeMismatch {
		t.Fatalf("arity: %v", err)
	}
}

func TestLimits(t *testing.T) {
	e := mustEngine(t, "a: Int", amino.WithMaxDepth(3), amino.WithMaxLength(20))
	if _, err := e.Parse("(((a > 1)))"); code(err) != amino.CodeDepthExceeded {
		t.Fatalf("depth: %v", err)
	}
	if _, err := e.Parse("a > 1 and a > 1 and a > 1"); code(err) != amino.CodeSyntax {
		t.Fatalf("length: %v", err)
	}
}

func TestOperatorConflict(t *testing.T) {
	_, err := amino.LoadSchema("a: Int", amino.WithOperator(amino.OperatorDef{Symbol: "=", BindingPower: 40}))
	if code(err) != amino.CodeOperatorConflict {
		t.Fatalf("got %v", err)
	}
}

File: typescript/test/engine.test.ts
import { describe, expect, it } from "vitest";

import { AminoError, loadSchema } from "../src/index.js";

function code(fn: () => unknown): string {
  try {
    fn();
  } catch (e) {
    if (e instanceof AminoError) return e.code;
    throw e;
  }
  return "";
}

describe("schema", () => {
  it("parses, indexes, and exports", () => {
    const engine = loadSchema(`# comment
struct Address { street: Str, city: Str }
struct Customer {
    id: Str
    billing: Address
    shipping: Address?
}
amount: Int {min: 0, max: 100}
tags: List[Str|Int]
email: Str?
customer: Customer
score: (a: Int, b: Float?) -> Float
`);
    const s = engine.schema;
    expect(s.structs).toHaveLength(2);
    expect(s.fields).toHaveLength(4);
    expect(s.functions).toHaveLength(1);
    expect(s.lookup("customer.billing.city")?.typeName).toBe("Str");
    expect(s.lookup("tags")?.elementTypes).toEqual(["Str", "Int"]);
    expect(s.lookup("amount")?.constraints).toEqual({ min: 0, max: 100 });
    expect(engine.exportSchema()).toBe(
      [
        "struct Address {street: Str, city: Str}",
        "struct Customer {id: Str, billing: Address, shipping: Address?}",
        "amount: Int {max: 100, min: 0}",
        "tags: List[Str|Int]",
        "email: Str?",
        "customer: Customer",
        "score: (a: Int, b: Float?) -> Float",
      ].join("\n"),
    );
  });

  it("rejects bad schemas", () => {
    for (const src of ["a: Nope", "a: Int\na: Str", "struct A { b: B }\nstruct B { a: A }", "struct S { x: Int, x: Str }", "a Int", "struct: Int"]) {
      const c = code(() => loadSchema(src));
      expect(["schema_parse", "schema_validation"], src).toContain(c);
    }
  });
});

describe("evaluation", () => {
  const engine = loadSchema("amount: Int\nstate: Str");
  const rules = [
    { id: "1", expr: "amount > 0 and state = 'CA'", meta: { ordering: 3 } },
    { id: "2", expr: "amount > 10 and state = 'CA'", meta: { ordering: 2 } },
    { id: "3", expr: "amount >= 100", meta: { ordering: 1 } },
  ];
  const rec = { id: 45, amount: 100, state: "CA" };

  it("all", () => {
    const r = engine.eval(rules, rec);
    expect(r.matched).toEqual(["1", "2", "3"]);
    expect(r.id).toBe(45);
  });
  it("first by ordering", () => {
    expect(engine.eval(rules, rec, { mode: "first", key: "ordering", order: "asc" }).matched).toEqual(["3"]);
    expect(engine.eval(rules, rec, { mode: "first", key: "ordering", order: "desc" }).matched).toEqual(["1"]);
  });
  it("inverse", () => {
    expect(engine.eval(rules, { amount: 5, state: "NY" }, { mode: "inverse" }).excluded).toEqual(["1", "2", "3"]);
  });
  it("score", () => {
    const r = engine.eval([{ id: "a", expr: "amount" }, { id: "b", expr: "amount > 0" }], rec, { mode: "score", threshold: 50 });
    expect(r.score).toBe(101);
    expect(r.matched).toEqual(["a", "b"]);
  });
  it("missing field makes the rule false, or tolerates a failing left side", () => {
    const e = loadSchema("a: Int\nb: Int?");
    const r = e.eval([{ id: "r", expr: "not b = 1" }, { id: "s", expr: "b = 1 or a = 1" }], { a: 1 });
    expect(r.matched).toEqual(["s"]);
  });
});

describe("validation", () => {
  it("strict mode throws on type, constraint, and custom type failures", () => {
    const strict = loadSchema("n: Int {min: 0}\ncontact: email", { decisionsMode: "strict" });
    const rule = [{ id: "r", expr: "n > 1" }];
    expect(code(() => strict.eval(rule, { n: "x", contact: "a@b.co" }))).toBe("decision_validation");
    expect(code(() => strict.eval(rule, { n: -1, contact: "a@b.co" }))).toBe("decision_validation");
    expect(code(() => strict.eval(rule, { n: 2, contact: "nope" }))).toBe("decision_validation");
  });
  it("loose mode drops the field with a warning", () => {
    const loose = loadSchema("n: Int\ncontact: email");
    const r = loose.eval([{ id: "r", expr: "contact = 'nope'" }], { n: 2, contact: "nope" });
    expect(r.matched).toEqual([]);
    expect(r.warnings).toHaveLength(1);
  });
});

describe("extensibility", () => {
  it("functions and custom operators", () => {
    const engine = loadSchema("text: Str\ntoxicity: (t: Str) -> Float\na: Int\nb: Int", {
      functions: { toxicity: () => 0.9 },
      customOperators: [
        {
          keyword: "near",
          bindingPower: 40,
          inputTypes: ["Int", "Int"],
          returnType: "Bool",
          fn: (x, y) => Math.abs((x as number) - (y as number)) < 5,
        },
      ],
    });
    const r = engine.eval([{ id: "t", expr: "toxicity(text) > 0.5" }, { id: "n", expr: "a near b" }], { text: "hi", a: 3, b: 6 });
    expect(r.matched).toEqual(["t", "n"]);
    expect(code(() => engine.parse("toxicity(a) > 0.5"))).toBe("type_mismatch");
    expect(code(() => engine.parse("toxicity()"))).toBe("type_mismatch");
  });
  it("limits", () => {
    const engine = loadSchema("a: Int", { maxDepth: 3, maxLength: 20 });
    expect(code(() => engine.parse("(((a > 1)))"))).toBe("depth_exceeded");
    expect(code(() => engine.parse("a > 1 and a > 1 and a > 1"))).toBe("syntax");
  });
  it("operator conflict", () => {
    expect(code(() => loadSchema("a: Int", { customOperators: [{ symbol: "=", bindingPower: 40 }] }))).toBe("operator_conflict");
  });
  it("validate() never throws for bad input", () => {
    const engine = loadSchema("a: Int");
    const r = engine.validate("a >");
    expect(r.ok).toBe(false);
    if (!r.ok) expect(r.error.code).toBe("syntax");
    expect(engine.validate("a > 1")).toEqual({ ok: true, type: "Bool" });
  });
});
